src/login.ts

```diff
@@ -6,52 +6,99 @@ import { MastoConfig } from './config';
 import { HttpNativeImpl } from './http';
 import type { LogType } from './logger';
 import { LoggerConsoleImpl } from './logger';
+import type { v1, v2 } from './mastodon';
 import { Client } from './mastodon';
-import { InstanceRepository } from './mastodon/v1/repositories';
+import { InstanceRepository as V1InstanceRepository } from './mastodon/v1/repositories';
+import { InstanceRepository as V2InstanceRepository } from './mastodon/v2/repositories';
 import { SerializerNativeImpl } from './serializers';
 import type { Writable } from './utils';
 import { WsNativeImpl } from './ws';
 
-export type LoginParams = {
+export interface RequestParams {
   /** URL of your instance */
   readonly url: string;
-  /** Log level to print to your console */
-  readonly logLevel?: LogType;
-  /** Access token of your account */
-  readonly accessToken?: string;
   /** Timeout in milliseconds */
   readonly timeout?: number;
   readonly defaultRequestInit?: Omit<RequestInit, 'body' | 'method'>;
-  readonly disableVersionCheck?: boolean;
   readonly disableDeprecatedWarning?: boolean;
-};
+}
 
-export const login = async (params: LoginParams): Promise<Client> => {
-  const draft: Writable<MastoConfigProps> = {
+const buildConfigProps = (params: RequestParams): MastoConfigProps => {
+  return {
     url: params.url,
     streamingApiUrl: '',
```

Comment:
I'd like to avoid adding `buildConfigProps` and `RequestParams` just for circumventing that `streamingApiUrl` can't be left undefined in the `MastoConfig` constructor. It looks pretty much the same as the `MastoProps` except that it doesn't accept `accessToken` `version` and some others.

It'd be better to add `?` to this line and make it work without `streamingApiUrl`:
https://github.com/neet/masto.js/blob/828671093b4c1db7e33826d6cb404de419c21ae1/src/config.ts#L20

I think it is pretty reasonable to accept `undefined` as `streamingApiUrl` because we now have `fetch{V1,V2}Instace` that does not require a WebSocket connection. It will come with some type errors, so you can throw errors to tell users about the lack of `streamingApiUrl`.

Reply:
Actually, `version`, `logLevel`, `disableVersionCheck` is unnecessary when using `fetch{V1,V2}Instace`. I don't think it's a good idea to require users to pass them but we never use them.

I think `streamingApiUrl` could be nullable too, but maybe we could do it in another PR

Reply:
> `version`, `logLevel`, `disableVersionCheck` is unnecessary when using `fetch{V1,V2}Instace`

I agree that we should not request users to specify unnecessary parameters, but I don't think it is the case.

If a Mastodon instance returns an invalid server (See #666 and #713), Masto.js will throw a parse error while checking the version. `disableVersionCheck` bypasses the version check for such a case and behaves as if it's running the latest version instead. Also, If you set `logLevel: "debug"` you can see the ongoing network request on your debug console.

As for `version`, I understood it's not needed. So apparently it is needed to declare different parameter types for each endpoint. This is what I'm thinking now:

| No.  | Name                                 | Required Parameters | Optional Parameters                                                                             |
| --- | ------------------------------------ | ------------------ | ---------------------------------------------------------------------------------------------- |
| \#1 | `fetchV1Instance`, `fetchV2Instance` | `url`              | `logLevel`, `timeout`, `defaultRequestInit`, `disableVersionCheck`, `disableDeprecatedWarning` |
| \#2 | `login`                              | `url`              | \#1 + `accessToken`                                                                          |
| \#3 | `createClient`                       | `url`              | \#2 + `streamingApiUrl`, `version`                                                           |

> I think streamingApiUrl could be nullable too, but maybe we could do it in another PR

Ok! Can we probably remove `buildConfigProps` after it? Because `RequestParams` is just a subset of `MastoConfigProps` except for `streamingApiUrl`'s nullability so we should be able to pass it directly to its constuctor.

Reply:
1. `disableVersionCheck` isn't used in `fetchV1/2Instance`. `version` is always `undefined`, so the `satisfiesVersion` function and `disableVersionCheck` is never used.
https://github.com/neet/masto.js/blob/a3da88cfa99f467e1ccbf350fbcdc3e0478f901b/src/config.ts#L108-L112

So I don't think it's a good idea to put `disableVersionCheck` to No.1. I prefer to keep it in No.2

2. I moved `logLevel` to No.1

3. `buildConfigProps` is used for building common fields & avoiding repeat code. Sure, we could remove this function, but what are the advantages of repeating code? 🤔

Reply:
> version is always undefined, so the satisfiesVersion function and disableVersionCheck is never used

Ah, I'm sorry you're right. Let's make it No. 2

> what are the advantages of repeating code?

I wonder what would we repeat after `streamingApiUrl` become nullable. `RequestParams extends MastoConfig` should be `true`, so I suppose we can

```diff
export const fetch{V1,V2}Instance = (
  params: RequestParams,
): Promise<v1.Instance> => {
- const { http, config } = buildHttpContext(buildConfigProps(params));
+ const { http, config } = buildHttpContext(params);
  return new V1InstanceRepository(http, config).fetch();
};
```

and

```diff
export const createClient = (params: LoginParams): Client => {
  const props: Writable<MastoConfigProps> = {
-   ...buildConfigProps(params),
+   ...params,
    streamingApiUrl: params.streamingApiUrl,
    logLevel: params.logLevel,
    accessToken: params.accessToken,
    disableVersionCheck: params.disableVersionCheck,
    version: params.version ? new SemVer(params.version) : undefined,
  };
  const { serializer, config, logger, http } = buildHttpContext(props);
```

or do you suppose we have any other logic that modifies params before passing it to MastoConfig? Just out of curiosity😄

Reply:
Oh, I must be stupid!
I just realized that `MastoConfigProps` and `LoginParams` are almost the same thing. I refactored and simplified the code. ❤️